modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+    This allows other concurrent, even mutating, operations while iterating,
+    but opens the possibility to iterating over duplicates or missing elements
+    from concurrent operations.
+
+    **Warning:** Breaking from an iterator will leak the snapshot due to issue #6912.
+    It will however leave all Bags in a safe state.
+  */
+  iter these() : eltType {
+    for loc in targetLocales {
+      for segmentIdx in 0..#here.maxTaskPar {
+        // The size of the snapshot is only known once we have the lock.
+        // BUG: c_ptr does not seem to work here and causes a segmentation fault at runtime.
+        var dom : domain(1) = {0..-1};
+        var buffer : [dom] eltType;
+
+        on loc {
+          ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+
+          if segment.acquireIfNonEmpty(STATUS_LOOKUP) {
+            var block = segment.headBlock;
+            while block != nil {
+              for i in 0 .. #block.size {
+                  buffer.push_back(block.elems[i]);
+              }
+              block = block.next;
+            }
+            segment.releaseStatus();
+
+          }
+        }
+
+        // Process this chunk if we have one...
+        for elem in buffer {
+          yield elem;
+        }
+      }
+    }
+  }
+
+  iter these(param tag : iterKind) where tag == iterKind.leader {
+    coforall loc in targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      coforall segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+
+        if segment.acquireIfNonEmpty(STATUS_LOOKUP) {
+          // Create a snapshot...
+          var block = segment.headBlock;
+          var bufferSz = segment.nElems.read() : int;
+          var buffer = c_malloc(eltType, bufferSz);
+          var bufferOffset = 0;
+
+          while block != nil {
+            if bufferOffset + block.size > bufferSz {
+              halt("DistributedBag Internal Error: Snapshot attempt with bufferSz(", bufferSz, ") with offset bufferOffset(", bufferOffset + block.size, ")");
+            }
+            __primitive("chpl_comm_array_put", block.elems[0], here.id, buffer[bufferOffset], block.size);
+            bufferOffset += block.size;
+            block = block.next;
+          }
+
+          // Yield this chunk to be process...
+          segment.releaseStatus();
+          yield (bufferSz, buffer);
+          c_free(buffer);
+        }
+      }
+    }
+  }
+
+  iter these(param tag : iterKind, followThis) where tag == iterKind.follower {
+    var (bufferSz, buffer) = followThis;
+    for i in 0 .. #bufferSz {
+      yield buffer[i];
+    }
+  }
+}
+
+/*
+  A segment block is an unrolled linked list node that holds a contiguous buffer
+  of memory. Each segment block size *should* be a power of two, as we increase the
+  size of each subsequent unroll block by twice the size. This is so that stealing
+  work is faster in that majority of elements are confined to one area.

```

Comment:
It's probably also worth pointing out here that any locking / parallel safety is taken care of outside of BagSegmentBlock.